Refactor `CallOutcome` to `Result`

MichaReiser opened the pull request:
## Summary

This PR refactors `Type::call` to better support the three main use cases:

* To check an **explicit call**, e.g. `a(1, 2)`: We want to warn about incorrect arguments.
* To check an **implicit call**, e.g. `a += 1`, `with a:`, `if a` (`__bool__`), `len(a)`: We want to test if the operation is supported or use a specific error message if the operation is not supported because the type doesn’t implement the expected Protocol or convention (“protocol”) but we don’t want to emit errors about missing or incorrect arguments
* To **query** the “best known” return type. This is, indirectly, used when evaluating visibility constraints (`__bool__`) but may also be used to test if a type implements a specific “protocol” or Protocol. We don’t want to emit any errors.


The existing implementation already supported checking explicit calls well. However, it did emit diagnostics for implicit calls when the arguments didn't match the expected signature. 

```py
class Manager:
    def __enter__() -> str:
        return "foo"

    def __exit__(self, exc_type, exc_value, traceback): ...

context_expr = Manager()

with context_expr as f:
    reveal_type(f) 
```

Red Knot emitted a rather confusing error about incorrect arguments before:

```
crates/red_knot_python_semantic/resources/mdtest/with/sync.md:55 unexpected error: [too-many-positional-arguments] "Too many positional arguments to function `__enter__`: expected 0, got 1"
```

My main finding is that `Type::call` should consider the call as failed if there are any errors, including binding errors. This is important for implicit call checking as demonstrated by the error above or when probing if a `Class` implements a specific dunder method. 

`Type::call` should return an `Err` if the call fails for any reason and leave the proper handling to the caller. However, it should retain enough information for the call site to create a useful diagnostic. There's a tension here because including very accurate information is expensive to collect and the work is wasted if the caller doesn't care about it (because it only wants to know if the call succeeded). I also think that collecting the appropriate information probably requires knowing the context in which the method was called. I don't have a concrete example but we may want to implement custom diagnostics for when an operator failed and I suspect that collecting the necessary information will be its own side-adventure (@BurntSushi told me that this is at least the case for Rustc). 

The main work of this PR is to change `Type::call`, `Type::dunder_call` and `Type::bound_call` to now return a `Result` and dealing with the fallout from downstream usages. I also had to implement a few workarounds where the new behavior is too accurate. E.g. I had to explicitly ignore binding errors in a few cases where we also did so before. We should tackle those TODOs in separate PRs (my goal was to preserve existing behavior in most places).

Returning a `Result` is now often less ergonomic on the call site than the old `outcome.return_type`. This is intentional because using `return_type` is often wrong. I want to make the decision about what should happen if the call fails explicit: Do we have to emit a diagnostic? Can we ignore certain errors? What's the best recovery logic for this specific inference? That's why it's now necessary to match the variants and call `Ok(outcome) => outcome.return_type(db)`,  `Err(err) => err.unwrap_return_type(db)`, or `Err(err) => err.return_type(db)` to get the return type best suited for this call site. 

## What's next

We should refactor our other `Outcome` types to use `Result`. We should explore whether we can move some methods that take a `context` and node into `infer.rs`. We should also consider whether `bool`, `len`, and `member` (and even `to_instance`) should return a `Result`,.

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -2039,88 +2039,54 @@ impl<'db> Type<'db> {
             Type::FunctionLiteral(function_type) => {
                 let mut binding = bind_call(db, arguments, function_type.signature(db), self);
                 match function_type.known(db) {
-                    Some(KnownFunction::RevealType) => {
-                        let revealed_ty = binding.one_parameter_type().unwrap_or(Type::unknown());
```
I moved this to `infer_call_expression` because those checks are only relevant when we perform an actual call (It would be surprising if an implicit call, e.g. from a `+=` emits a revealed diagnostic.

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -2153,108 +2115,153 @@ impl<'db> Type<'db> {
+                let mut bindings = Vec::with_capacity(elements.len());
+                let mut not_callable = Vec::new();
+
+                for element in elements {
```
I don't feel to confident about the unrolling of the inner call result here.

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -2310,15 +2368,14 @@ impl<'db> Type<'db> {
+            Symbol::Type(callable_ty, Boundness::Bound) => Ok(callable_ty.call(db, arguments)?),
             Symbol::Type(callable_ty, Boundness::PossiblyUnbound) => {
-                CallDunderResult::PossiblyUnbound(callable_ty.call(db, arguments))
+                let call = callable_ty.call(db, arguments)?;
+                Err(CallDunderError::PossiblyUnbound(call))
```
This is a small change but a not callable type error now takes precedence over a possibly unbound error.

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -4149,27 +4235,30 @@ impl<'db> Class<'db> {
+                    kind: MetaclassErrorKind::PartlyNotCallable(metaclass),
                 }),
 
                 // TODO we should also check for binding errors that would indicate the metaclass
                 // does not accept the right arguments
```
This should be easy now ;)

MichaReiser commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -1618,7 +1621,7 @@ impl<'db> TypeInferenceBuilder<'db> {
                 let target_ty = enter_ty
                     .call(self.db(), &CallArguments::positional([context_expression_ty]))
-                    .return_type_result(&self.context, context_expression.into())
+                    .map(|outcome| outcome.return_type(self.db()))
```
We probably want to have more fine granular error handlers here but that's out of the scope of this PR (and the same applies for other implicit call sites).

carljm replied:
Worth a TODO comment?

MichaReiser commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -5939,23 +6095,27 @@ fn perform_rich_comparison<'db>(
-        }
-    };
+    // TODO: This should use `call_dunder` instead of reimplementing it from scratch.
+    let call_dunder =
```
We can't use `call_dunder` here because `call_dunder` contains a `to_meta_class` call. Maybe we should consider removing it and let the call site deal with converting to a meta class first.

carljm replied:
I don't think we should do that, because "looked up on the class, not the instance" is part of the contract of calling a dunder method in general in Python. And it is true here as well: see the use of `left.class.class_member` below, which is equivalent (for an Instance type) on looking up a member on the meta type.

AlexWaygood reviewed (commented):
A review of the tests. Haven't looked at the code yet ;)

MichaReiser commented on crates/red_knot_python_semantic/resources/mdtest/loops/for.md:
```diff
@@ -247,7 +247,7 @@ class Test2:
 def _(flag: bool):
     # error: "Object of type `Test | Test2` is not iterable"
```
Yes, we currently only handle the `PossiblyUnbound` error case but we don't handle unions at all (other than saying you can't call it)

MichaReiser commented on crates/red_knot_python_semantic/resources/mdtest/with/sync.md:
```diff
@@ -134,3 +134,20 @@ def _(flag: bool):
+
+context_expr = Manager()
+
+# error: [invalid-context-manager] "Object of type `Manager` cannot be used with `with` because the method `__enter__` of type `Literal[__enter__]` is not callable"
```
That's fair. It requires improving the context manager error messages. I can add a TODO that we need to remove the catch all error handler and instead handle each error manually

AlexWaygood replied:
SGTM

MichaReiser commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -3249,9 +3249,169 @@ impl<'db> TypeInferenceBuilder<'db> {
+        match call {
+            Ok(outcome) => {
+                for binding in outcome.bindings() {
+                    let Some(known_function) = binding
```
@carljm you might be disappointed by this but I don't think the diagnostics should be part of `CallOutcome` because they are only relevant if we perform an actual function call. Implicit calls want to use their own custom diagnostics, so it doesn't make sense to expose this call expression specific behavior on the generic `CallOutcome`.

carljm replied:
> they are only relevant if we perform an actual function call. Implicit calls want to use their own custom diagnostics

I'm just not convinced that this is universally true. I still think we will probably need some form of nested or chained diagnostics to provide adequate detailed debugging information to the user in a case where something like (for example) an "object is not callable" error is raised as part of an implicit call, e.g. because some non-callable object has been used as the value of some dunder. So I do expect that we will need the equivalent of some of the below diagnostics to be rendered as part of some errors on implicit calls. But it's possible that this could be done just by extracting the necessary diagnostics into standalone functions in `diagnostics.rs` and calling them from multiple callsites.

MichaReiser replied:
That's possible. Although it comes with its own challenges because we don't have a call node. So what do you pass as the node that the highlighting works as expected? Either way. My follow up PRs introduce a pattern where the `Err` type has methods to emit the corresponding diagnostics and get the return type.

MichaReiser commented:
I now updated past @carljm changes regarding `float`, `int` and `complex` handling and removed the related hacks. However, there are now a few failing tests because we fail to bind the calls:

```
failures:

---- mdtest__binary_instances stdout ----

instances.md - Binary operations on instances - Callable instances as dunders

  crates/red_knot_python_semantic/resources/mdtest/binary/instances.md:263 unmatched assertion: revealed: Unknown | int
  crates/red_knot_python_semantic/resources/mdtest/binary/instances.md:263 unexpected error: 13 [unsupported-operator] "Operator `+` is unsupported between objects of type `B` and `B`"
  crates/red_knot_python_semantic/resources/mdtest/binary/instances.md:263 unexpected error: 1 [revealed-type] "Revealed type is `Unknown`"

To rerun this specific test, set the environment variable: MDTEST_TEST_FILTER="instances.md - Binary operations on instances - Callable instances as dunders"
MDTEST_TEST_FILTER="instances.md - Binary operations on instances - Callable instances as dunders" cargo test -p red_knot_python_semantic --test mdtest -- mdtest__binary_instances

--------------------------------------------------

thread 'mdtest__binary_instances' panicked at crates/red_knot_test/src/lib.rs:98:5:
Some tests failed.
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- mdtest__comparison_integers stdout ----

integers.md - Comparison: Integers - Integer literals

  crates/red_knot_python_semantic/resources/mdtest/comparison/integers.md:16 unmatched assertion: revealed: Unknown
  crates/red_knot_python_semantic/resources/mdtest/comparison/integers.md:16 unexpected error: 1 [revealed-type] "Revealed type is `Unknown & ~AlwaysTruthy | Literal[True]`"

To rerun this specific test, set the environment variable: MDTEST_TEST_FILTER="integers.md - Comparison: Integers - Integer literals"
MDTEST_TEST_FILTER="integers.md - Comparison: Integers - Integer literals" cargo test -p red_knot_python_semantic --test mdtest -- mdtest__comparison_integers

--------------------------------------------------

thread 'mdtest__comparison_integers' panicked at crates/red_knot_test/src/lib.rs:98:5:
Some tests failed.

---- mdtest__comparison_instances_rich_comparison stdout ----

rich_comparison.md - Comparison: Rich Comparison - Numbers Comparison with typeshed

  crates/red_knot_python_semantic/resources/mdtest/comparison/instances/rich_comparison.md:310 unmatched assertion: revealed: Unkown
  crates/red_knot_python_semantic/resources/mdtest/comparison/instances/rich_comparison.md:310 unexpected error: 1 [revealed-type] "Revealed type is `Unknown`"

To rerun this specific test, set the environment variable: MDTEST_TEST_FILTER="rich_comparison.md - Comparison: Rich Comparison - Numbers Comparison with typeshed"
MDTEST_TEST_FILTER="rich_comparison.md - Comparison: Rich Comparison - Numbers Comparison with typeshed" cargo test -p red_knot_python_semantic --test mdtest -- mdtest__comparison_instances_rich_comparison

--------------------------------------------------

thread 'mdtest__comparison_instances_rich_comparison' panicked at crates/red_knot_test/src/lib.rs:98:5:
Some tests failed.

---- mdtest__comparison_non_bool_returns stdout ----

non_bool_returns.md - Comparison: Non boolean returns

  crates/red_knot_python_semantic/resources/mdtest/comparison/non_bool_returns.md:36 unexpected error: [unsupported-operator] "Operator `<` is not supported for types `int` and `A`, in comparing `Literal[1]` with `A`"
  crates/red_knot_python_semantic/resources/mdtest/comparison/non_bool_returns.md:37 unmatched assertion: revealed: Literal[False] | A
  crates/red_knot_python_semantic/resources/mdtest/comparison/non_bool_returns.md:37 unexpected error: 1 [revealed-type] "Revealed type is `Unknown & ~AlwaysTruthy | A`"
  crates/red_knot_python_semantic/resources/mdtest/comparison/non_bool_returns.md:39 unexpected error: [unsupported-operator] "Operator `<` is not supported for types `int` and `A`, in comparing `Literal[0]` with `A`"

To rerun this specific test, set the environment variable: MDTEST_TEST_FILTER="non_bool_returns.md - Comparison: Non boolean returns"
MDTEST_TEST_FILTER="non_bool_returns.md - Comparison: Non boolean returns" cargo test -p red_knot_python_semantic --test mdtest -- mdtest__comparison_non_bool_returns

--------------------------------------------------

thread 'mdtest__comparison_non_bool_returns' panicked at crates/red_knot_test/src/lib.rs:98:5:
Some tests failed.

---- mdtest__comparison_intersections stdout ----

intersections.md - Comparison: Intersections - Positive contributions

  crates/red_knot_python_semantic/resources/mdtest/comparison/intersections.md:30 unmatched assertion: revealed: bool
  crates/red_knot_python_semantic/resources/mdtest/comparison/intersections.md:30 unexpected error: 25 [unsupported-operator] "Operator `>` is not supported for types `Child1` and `int`, in comparing `Child1 & Child2` with `Literal[1]`"
  crates/red_knot_python_semantic/resources/mdtest/comparison/intersections.md:30 unexpected error: 13 [revealed-type] "Revealed type is `Unknown`"

To rerun this specific test, set the environment variable: MDTEST_TEST_FILTER="intersections.md - Comparison: Intersections - Positive contributions"
MDTEST_TEST_FILTER="intersections.md - Comparison: Intersections - Positive contributions" cargo test -p red_knot_python_semantic --test mdtest -- mdtest__comparison_intersections

--------------------------------------------------

thread 'mdtest__comparison_intersections' panicked at crates/red_knot_test/src/lib.rs:98:5:
Some tests failed.


failures:
    mdtest__binary_instances
    mdtest__comparison_instances_rich_comparison
    mdtest__comparison_integers
    mdtest__comparison_intersections
    mdtest__comparison_non_bool_returns
```


The pattern for all tests is roughly the same: 

```
from __future__ import annotations

class A:
    def __lt__(self, other) -> A: ...

class B:
    def __lt__(self, other) -> B: ...

class C:
    def __lt__(self, other) -> C: ...

x = A() < B() < C()
reveal_type(x)  # revealed: A & ~AlwaysTruthy | B

y = 0 < 1 < A() < 3
reveal_type(y)  # revealed: Literal[False] | A

z = 10 < 0 < A() < B() < C()
reveal_type(z)  # revealed: Literal[False]
```

Binding the call results in a todo return type... 

```
return_ty: Dynamic(
        Todo(
            Message(
                "Unsupported or invalid type in a type expression",
            ),
        ),
    ),
```

MichaReiser commented on crates/red_knot_python_semantic/resources/mdtest/comparison/intersections.md:
```diff
@@ -14,7 +14,12 @@ class Child1(Base):
         return True
 
-class Child2(Base): ...
+    def __gt__(self, other) -> bool:
```
@sharkdp I'm pretty sure I now changed the test in a way that it is no longer testing what it should but I'm not sure how to fix it without breaking the test itself.

sharkdp replied:
> I think you should put this method on `Base` instead of `Child1` and `Child2`.

Yes. Reverting to the state on main and adding `__get__` to `Base` keeps the original intention of the test and fixes the error. Thanks!

MichaReiser commented on crates/red_knot_python_semantic/resources/mdtest/comparison/non_bool_returns.md:
```diff
@@ -23,6 +23,7 @@ from __future__ import annotations
 
 class A:
     def __lt__(self, other) -> A: ...
+    def __gt__(self, other) -> A: ...
```
@sharkdp Same here. I'm not sure if the fix is in the "spirit of the test" but it was wrong before ;)

MichaReiser commented on crates/red_knot_python_semantic/resources/mdtest/binary/instances.md:
```diff
@@ -263,7 +260,10 @@ class B:
 
 # TODO: this could be `int` if we declare `B.__add__` using a `Callable` type
-reveal_type(B() + B())  # revealed: Unknown | int
+# TODO: This should not be an error but we currently fail to bind `self` in `A.__call__` as `B`
```
This is a "regression" in the sense that we need to understand that `__add__ = A` rebinds the `self` in `A.__call__` to `B`.

MichaReiser commented:
I'm pretty happy with the test improvements :)

carljm reviewed (commented):
I've only reviewed the tests so far, not the implementation. Submitting this just in case I don't make it through the implementation before I'm done for the day. It's great to see all those operator TODOs removed!

AlexWaygood commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -2156,108 +2115,100 @@ impl<'db> Type<'db> {
                     }
 
-                    _ => CallOutcome::callable(binding),
+                    _ => {}
```
I think I'd prefer an exhaustive match here, since it seems like we special-case the return-type for a majority of the `KnownFunction` variants (at least right now):

```suggestion
                    // These remaining functions do not have the inferred return-type special-cased in any way.
                    Some(
                        KnownFunction::ConstraintFunction(_)
                        | KnownFunction::Final
                        | KnownFunction::RevealType
                        | KnownFunction::NoTypeCheck
                        | KnownFunction::StaticAssert
                        | KnownFunction::AssertType,
                    )
                    | None => {}
```

AlexWaygood commented on crates/red_knot_python_semantic/src/types/call.rs:
```diff
@@ -1,423 +1,262 @@
+    /// dunder is possibly unbound).
+    ///
+    /// If the type is not callable, returns `Type::Unknown`.
+    pub(super) fn unwrap_return_type(&self, db: &'db dyn Db) -> Type<'db> {
+        self.return_type(db).unwrap_or(Type::unknown())
+    }
```
I think I'd expect a function with `unwrap` in its name to panic if it contained an error. But this returns a fallback in the case of an error. And in fact, we know we have an error here, since this is a method on `CallError` 😄 Maybe something like

```suggestion
    pub(super) fn fallback_return_type(&self, db: &'db dyn Db) -> Type<'db> {
        self.return_type(db).unwrap_or(Type::unknown())
    }
```

?

AlexWaygood commented on crates/red_knot_python_semantic/src/types/call.rs:
```diff
@@ -1,423 +1,262 @@
-            }
-        }
-    }
+    /// Tries calling each union element using the provided `call` function.
```
```suggestion
    /// Constructor that tries calling each union element using the provided `call` function.
```

AlexWaygood commented on crates/red_knot_python_semantic/src/types/call.rs:
```diff
@@ -1,423 +1,262 @@
-                            node,
-                            format_args!("Static assertion error: {message}"),
-                        );
+                    // Should this be OK, or an error? We can't make it not_callable
+                    // because calling it would actually work because it ignores the
+                    // possibly unboundness.
+                    CallError::PossiblyUnboundDunderCall { outcome, .. } => match *outcome {
+                        CallOutcome::Union(inner_bindings) => {
+                            bindings.extend(inner_bindings);
+                        }
+                        CallOutcome::Single(binding) => {
+                            bindings.push(binding);
+                        }
+                    },
```
I'm having trouble grappling with this question. Can you come up with a Python snippet where we'd change behaviour if we did this branch differently? It would help me think about the issue more clearly, I think

MichaReiser replied:
I think the example here would be:

```py
def outer(cond1: bool):
	class Test:
		if cond:
			def __call__(self): ...
	
	class Other:
		def __call__(self): ...
	
	def inner(cond: bool): 
		if cond:
			a = Test()
		else:
			a = Other()
	
		a()
```

`a` resolves to `Test | Other`. Calling `Test` resolves to `PossiblyUnbound` and calling `Other` to `Ok` (let's ignore whether the `__call__` is actually correct). 

I do think what I have now is incorrect and I should push it to the error variant instead.

MichaReiser replied:
I added a test case for this (with a TODO to improve the error message)

AlexWaygood reviewed (approved):
This looks excellent overall. There's definitely still a few outstanding issues but I'm overall leaning towards adding TODOs for the issues and landing this, then iterating. It's a big PR that is liable to accumulate merge conflicts, and it definitely seems like a big improvement over what we have now

MichaReiser commented on crates/red_knot_python_semantic/src/types/call.rs:
```diff
@@ -1,423 +1,262 @@
-    CustomError(&'db str),
-}
-
+/// A successfully bound call where all arguments are valid.
```
I decided not to split `CallOutcome` into two versions where one is statically known to not have any errors and one that has. Instead, guaranteeing that the `Ok` variant never contains errors is left to `Type::call`. I haven't explored how complicated it would be to split `CallBinding` but it is something we could look into as a follow up (e.g. `bind_call` could return a `Result` too? Although we still want to get a binding even if there are errors)

MichaReiser replied:
I'm not sure what you mean. We only return `CallOutcome `if there are no binding errors (all arguments are valid).

MichaReiser replied:
As discussed async. @AlexWaygood and I will follow up on this after landing this PR (if it still is an issue)

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -2156,108 +2115,100 @@ impl<'db> Type<'db> {
+                            .map(|arg| arg.bool(db).into_type(db))
+                            .unwrap_or(Type::BooleanLiteral(false)),
+
+                        Some(KnownClass::Str) => arguments
+                            .first_argument()
+                            .map(|arg| arg.str(db))
+                            .unwrap_or(Type::string_literal(db, "")),
```
Yes. It's something that also comes up with `bool`. But it mainly comes down that we need to bind class constructor calls, which we currently don't. We should then get the signature check for free?

MichaReiser replied:
That's fair. So we might need to do more but we also need to check that the argument passed to bool is correct (there's exactly one and it implements the bool convention)

AlexWaygood replied:
> I also only moved this code. So this isn't new

I know, I said it was probably a pre-existing issue 😄 but it would still be good to add a TODO while we're here, and while I've noticed it

MichaReiser replied:
Absolutely. It wasn't my intention to suggest otherwise.

carljm commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -2267,13 +2218,12 @@ impl<'db> Type<'db> {
         receiver_ty: &Type<'db>,
         arguments: &CallArguments<'_, 'db>,
-    ) -> CallOutcome<'db> {
+    ) -> Result<CallOutcome<'db>, CallError<'db>> {
```
Is it worth aliasing this type as `CallResult`?

MichaReiser replied:
I prefer repeating types because I find it easier to read the code unless the type is used in many places, which isn't the case here (I think it's only three places today and David's about to remove `call_bound`, leaving only 2o

carljm commented on crates/red_knot_python_semantic/src/types/infer.rs:
```diff
@@ -6000,7 +6149,14 @@ fn perform_membership_test_comparison<'db>(
                     &CallArguments::positional([Type::Instance(right), Type::Instance(left)]),
                 )
-                .return_type(db)
+                // TODO: We shouldn't ignore binding errors here but we have to because we
+                // currently don't support assigning ints to floats.
```
Seems like this should have been fixed by my PR?

MichaReiser replied:
ah nice catch. This must have gotton lost somewhere.

MichaReiser commented on crates/red_knot_python_semantic/src/types/call.rs:
```diff
@@ -1,423 +1,262 @@
+                        }
+                    },
+                    CallError::BindingError { binding } => {
+                        not_callable.push(NotCallableVariant::new(
```
Do you have any suggestions? It needs to be something generic because we otherwise end up with too many error variants that bloat the size, are hard to handle, and upstream code doesn't seem to care anyway.

MichaReiser replied:
Let's merge this into your comment below

MichaReiser commented on crates/red_knot_python_semantic/resources/mdtest/call/union.md:
```diff
@@ -75,3 +75,36 @@ def _(flag: bool):
+    else:
+        f = f2
+
+    # error: [call-non-callable] "Object of type `Literal[f1, f2]` is not callable (due to union element `Literal[f2]`)"
```
I'd prefer improving this for now because I get the impression that we may also want to rename the lint rule itself to something more generic? But I could see how this itself requires some discussion.

MichaReiser commented on crates/red_knot_python_semantic/src/types/call.rs:
```diff
@@ -1,423 +1,262 @@
     }
 }
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(super) struct NotCallableVariant<'db> {
+    /// The resolved binding for that variant or `None` if the variant is not callable.
+    binding: Option<CallBinding<'db>>,
+
+    /// The variant's type that is not callable.
+    not_callable: Type<'db>,
+}
```
I deleted this type

MichaReiser commented:
Ha, less code, more tests and more accurate!

MichaReiser commented on crates/red_knot_python_semantic/src/types.rs:
```diff
@@ -1658,108 +1617,111 @@ impl<'db> Type<'db> {
-                            not_callable_ty: self,
+                instance_ty
+                    .call_dunder(db, "__call__", &arguments.with_self(instance_ty))
+                    .map_err(|err| match err {
+                        CallDunderError::Call(CallError::NotCallable { .. }) => {
+                            // Turn "`<type of illegal '__call__'>` not callable" into
+                            // "`X` not callable"
+                            CallError::NotCallable {
+                                not_callable_ty: self,
+                            }
                         }
```
This patching of the `not_callable_ty` here seems hacky but we should address this separately

MichaReiser commented:
Thank you both for the excellent review. 

I'm sorry for deferring so many diagnostic improvements but I see my priority as solving an architectural problem and I should then get back to working on the CLI. I added the requested todos and pushed some improvements to the `NotCallable` that hopefully addresses all terminology concerns. 

There's one behavior change that I want to call out. We now only report the first error if a union variant isn't callable or has binding errors, similar to what TypeScript does (It's no longer "X isn't callable (because of element reason)" but is now "element isn't callable"). This improved some errors but I think we can do better. I added a todo in code.

carljm commented on crates/red_knot_python_semantic/resources/mdtest/binary/integers.md:
```diff
@@ -10,9 +10,10 @@ reveal_type(-3 // 3)  # revealed: Literal[-1]
-# TODO: We don't currently verify that the actual parameter to int.__add__ matches the declared
-# formal parameter type.
-reveal_type(2 + "f")  # revealed: int
+# TODO: This should emit an unsupported-operator error but we don't currently
+#  verify that the actual parameter to `int.__add__` matches the declared
+#  formal parameter type.
```
I think this comment is now wrong, not in terms of what should happen (yes, there should be an unsupported-operator error), but the why. What is happening here after this PR is that we are falling back to `str.__radd__` (because we have the wrong argument type for `int.__add__` -- we _do_ now "verify that the actual parameter to `int.__add__` matches the declared formal parameter type"), but `str.__radd__` is overloaded and we don't understand that, so we fail to detect that this operation is also invalid for `str.__radd__`.

carljm replied:
I'll put up a quick PR for this; it's minor but I don't want to waste the time I spent understanding what's happening.

carljm replied:
https://github.com/astral-sh/ruff/pull/16242

carljm commented on crates/red_knot_python_semantic/resources/mdtest/call/callable_instance.md:
```diff
@@ -52,7 +52,7 @@ class NonCallable:
 
 a = NonCallable()
-# error: "Object of type `Unknown | Literal[1]` is not callable (due to union element `Literal[1]`)"
+# error: [call-non-callable] "Object of type `Literal[1]` is not callable"
```
This seems to me like clearly a regression, not an improvement; it's not clear to me why this change was made. Are there other error messages where it resulted in an improvement?

In any case, we can certainly follow up on this later; I can create an issue for it.

carljm replied:
Ah, ok, I suspect it was made to improve cases where we need more detailed errors from "inside" the union (like argument-specific call binding errors). That makes sense, but I continue to feel like in the end we won't be able to get away with choosing just one or the other here (either the union-level info or the per-union-element details), we will ultimately need to be able to chain details.

MichaReiser replied:
It's probably worth tagging @BurntSushi because this might be a use case where we want sub-diagnostics (although we still need a way to truncate to avoid going too deep). 

@BurntSushi the diagnostic use case we have here is that we want to warn if a user tries to call a type that can't be called with the given arguments. The part where our current system (and what I did in this PR) falls short is if the called type is a union where only one of the variants can't be called (because it isn't a function or the arguments don't match its signature). 

I did look at how different type checker handle this case when I started working on this refactor:

**Pyright**

```py
from ast import Call
from typing import reveal_type


class NotCallable:
    pass

class Callable:
    def __call__(self):
        pass


def coinflip() -> NotCallable | Callable:
    return random.choice([NotCallable(), Callable()])

a = coinflip()

reveal_type(a)

a()
```

It only reports an error for the not callable variant (similar to what I did in this PR). It does have some form of sub diagnostic.

```
Object of type "NotCallable" is not callable
  Attribute "__call__" is unknownPylancereportCallIssue
```

**Typescript**

```ts
class NonCallable {

}

class AlsoNotCallable {

}


let x: NonCallable | AlsoNotCallable | (<T>(a: T) => T) | undefined = a => a;

x("test")
```

```
    This expression is not callable.
      Not all constituents of type 'NonCallable | AlsoNotCallable | (<T>(a: T) => T)' are callable.
        Type 'NonCallable' has no call signatures.
```

Similar to `Pyright` where it uses sub-diagnostics.

But the quality of errors depends on the use case. E.g. there's no special union handling for iterable:


```ts
class MyIterable implements Iterable<number>  {
    [Symbol.iterator](): Iterator<number> {
        return [1, 2, 3, 4]
    }
}



class NotIterable {

}

let a:MyIterable | NotIterable= new MyIterable();

for (let i of a) {
    console.log(i);
}
```

```
Type 'MyIterable | NotIterable' must have a '[Symbol.iterator]()' method that returns an iterator.ts(2488)
```

It doesn't tell you that `NotIterable` is the problematic element.

**Flow**

Same examples as for TypeScript

```
    12: x("test")
        ^ Cannot call `x` because undefined [1] is not a function. [not-a-function]
        References:
        10: let x: NonCallable| (<T>(a: T) => T) | undefined = undefined;
                                                   ^ [1]
    12: x("test")
        ^ Cannot call `x` because a call signature declaring the expected parameter / return type is missing in `NonCallable` [1]. [prop-missing]
        References:
        10: let x: NonCallable| (<T>(a: T) => T) | undefined = undefined;
                   ^ [1]
```

```
15: for (let i of a) {
                  ^ property `@@iterator` is missing in `NotIterable` [1] but exists in `$Iterable` [2]. [prop-missing]

    References:

    13: let a:MyIterable | NotIterable= new MyIterable();
                           ^ [1]

    [LIB] ..//try-flow-website-js@0.260.0/flowlib/core.js:1865: interface $Iterable<+Yield,+Return,-Next>{                                                                      ^ [2]
    
```

carljm replied:
I filed https://github.com/astral-sh/ruff/issues/16241 just so we don't lose track of this discussion (and in particular this excellent comment which I linked to from the issue)

carljm reviewed (commented):
The updates look great, thank you! Reads much clearer to me now. And really appreciate all the TODOs for things we want to try to further improve.

MichaReiser commented on crates/red_knot_python_semantic/src/types/call.rs:
```diff
@@ -1,423 +1,206 @@
-            asserted_ty,
+        if errors.is_empty() {
+            Ok(CallOutcome::Union(bindings.into()))
+        } else if bindings.is_empty() && not_callable {
```
Line 47 is important.

You could have a case where one variant is `NotCallable` and the other variant has a `BindingError`. We don't want to return `NotCallable` in this case because it would remove information.

carljm replied:
But I can remove the `not_callable` boolean, and every reference to it (including line 47), and all tests still pass. So if it's important, either we aren't currently handling it correctly or we are missing some needed test.

In the case you describe, `bindings` will not be empty, so we won't error this clause anyway; the `not_callable` boolean is irrelevant.

MichaReiser replied:
I don't think `bindings` will be empty because we only push to `bindings` if the call succeeded (without any binding errors)

carljm replied:
Oh, I see! I missed that we don't push bindings with errors to `bindings`. Makes sense, I agree then it just looks like we don't have a test for the case where one union variant has a binding error and the other one is not callable (or we do have a test for that case but its assertions fail to distinguish?)
